processing/data_collection/gazette/spiders/base.py

```diff
@@ -18,6 +20,88 @@ def __init__(self, start_date=None, *args, **kwargs):
                 self.start_date = parsed_data.date()
 
 
+class DiarioMunicipalSigpubGazetteSpider(BaseGazetteSpider):
+    """www.diariomunicipal.com.br (Sigpub) base spider
+
+    Documents obtained by this kind of spider are text-PDFs with many cities in it.
+    That's because the websites are usually made for associations of cities.
+    
+    TODO:
+        - All variations have a "possible" start date of 01/01/2009, but that may cause
+        many unnecessary requests to be made if they actually start making available
+        documents later. Some investigation for the start date of each website needs to
+        be made in this case.
+
+    Observations:
+        - These websites have an "Advanced Search", but they are protected by ReCaptcha.
+    """
+
+    custom_settings = {"USER_AGENT": UserAgent().random}
+    FIRST_AVAILABLE_DATE = date(2009, 1, 1)
+
+    def start_requests(self):
+        """Requests start page where the calendar widget is available."""
+        yield scrapy.Request(self.CALENDAR_URL, callback=self.parse_calendar)
```

Comment:
The URLs are all the same, the only change in it is the association ID, it would look cleaner in the spiders that are using this base if they just pass association ID.

Example `url = f"{self.BASE_URL}/{self.ASSOCIATION_ID}/materia/calendario"` instead of `CALENDAR_URL`

And set in base class:

`BASE_URL = "http://www.diariomunicipal.com.br"`

And for each spider set:

`ASSOCIATION_ID = "somename"`

Reply:
Just some context:

When I was doing this I didn't know how every website would behave since I did them one at a time. So separating every endpoint to it's own variable seemed safe. At the end, the pattern holds for every one of them.

I agree we could repurpose some parts of the URL. But I think that at least `CALENDAR_URL` shouldn't be split for documentation purposes. It points directly to the spider website and having the entire URL is nice.

I'll repurpose the other ones. Thanks!

Reply:
Yeah, it does make sense, but the one from SC follows the approach I mentioned, still, looks cleaner than before.